### DISCUSSION TITLE
Using Material for MkDocs and PyMdownX for some REST API Doc design

### DISCUSSION THREAD

**Discussion by Andre601** [squidfunk/mkdocs-material#3909] [Show and tell]
The goal was to make API documentation that **MIMICS** OpenAPI in terms of design. With that said is the shown content NOT the result of using OpenAPI. This is all hand made using the `details` and `snippets` expansion from PyMdownX.

General look:
![image](https://user-images.githubusercontent.com/11576465/167728366-a14dabd7-3325-49b8-8c96-ffb6cb3fefc2.png)

POST content:
![image](https://user-images.githubusercontent.com/11576465/167728436-175f1236-362f-4079-b0f3-79709063ff99.png)
![image](https://user-images.githubusercontent.com/11576465/167728492-b8e74992-83ab-4105-ae0e-d1a5fbb8950f.png)

Success Response:
![image](https://user-images.githubusercontent.com/11576465/167728611-adc23634-484f-46fe-9f3c-606c88b1fd66.png)

Failed Responses:
![image](https://user-images.githubusercontent.com/11576465/167728662-5c131f15-064d-4914-b36f-e510a51bd7fc.png)
![image](https://user-images.githubusercontent.com/11576465/167728706-8523f086-7fb2-443a-9a79-d6bb79c500c0.png)

----

GET content:
![image](https://user-images.githubusercontent.com/11576465/167728734-c8b65056-7c00-4e42-a993-4db08d9b2094.png)

Success Response:
![image](https://user-images.githubusercontent.com/11576465/167728775-26ba6f98-157e-4b37-ab8c-d3ba7d7f4eb8.png)

Failed Responses:
![image](https://user-images.githubusercontent.com/11576465/167728826-7beb76aa-eb83-423e-8c04-760123bcfcac.png)
![image](https://user-images.githubusercontent.com/11576465/167728868-baf0f360-4864-4dc6-ada0-50f48227463b.png)

**Comment by squidfunk** [comment]
Very cool! I've seen some attempts to generate documentation from OpenAPI specs – really awesome to see some innovation in that space, and this already looks pretty good.

**Reply by Andre601** [reply]
Well... thing is that this isn't generated from OpenAPI secs. It's hand made.

**Reply by martinbira** [reply]
Ah :( I was excited because our implementation, https://docs.apptus.com/esales-4/integration/api/specification/ , is generated from our OpenAPI spec, but since we use Swagger UI, it looks like meh.

**Comment by jromal** [comment]
I used this for a test and it looks nice.
https://github.com/bharel/mkdocs-render-swagger-plugin

It connects online so it renders (almost) identical as on the swagger.io (some colors are not the same).

**Reply by Andre601** [reply]
What does the output look like?
I saw that plugin too, but the fact that there isn't even an example of how it would look like makes me not that interested in spending time to work with it...

**Comment by jromal** [comment]
It looks like the same as on the web. Actually it connects to the swagger.io or you need to download the swagger-ui.js file (like I do).

The only difference is that the combination of colors make it unreadable in some cases (e.g. examples with light option or some texts with dark option).


![image](https://user-images.githubusercontent.com/79361136/168529886-9b876939-9799-4646-819a-8d28490f781f.png)

![image](https://user-images.githubusercontent.com/79361136/168529946-5fd7f397-09bc-48e6-97f9-7c7b99f1ff61.png)

While the same at [swagger.io](https://editor.swagger.io/) looks much better because background is black. I guess, but could be totally wrong, that some of the settings of the material mkdocs overwrite the color of this background. But no idea at all.

![image](https://user-images.githubusercontent.com/79361136/168530464-3f1dae24-8134-402b-b73e-572f770e8c81.png)

**Reply by jromal** [reply]
A colleague of mine is using with a local server and using standard template colors. The background of examples is light grey and makes green not readable, and the dark mode makes all black text unreadable. 

Definitively there is a conflict between this plug-in and swagger-ui.js with Material for Mkdocs. Maybe some specific background colors can be forced for that javascript part. In any case it is not my biggest priority and will wait if/when someone finds a way to make it nicer.

Thanks.

**Reply by blueswen** [reply]
Hi @javier-romera. If you are still looking for a dark mode of Swagger UI compatible with Material for MkDocs, you can consider my  mkdocs plugin [mkdocs-swagger-ui-tag](https://github.com/blueswen/mkdocs-swagger-ui-tag). The effect as follow:

![Dark mode example](https://user-images.githubusercontent.com/1564148/173606676-7e219638-c686-474d-b2b4-715444a5084a.png)

**Comment by jromal** [comment]
Great. 

It solves my problem of lack of contrast in the examples that I had with the previous plugin.
Very nice job. 

Many thanks.

**Reply by Andre601** [reply]
You should reply within the answer and not in a separate answer... Makes it feel disconnected.

**Comment by Andre601** [comment]
The latest admonition changes look really nice for my API Docs (Again, no swagger, only manual design)
![SmartSelect_20221003-105339_Chrome.jpg](https://user-images.githubusercontent.com/11576465/193538790-6fae5d48-6f78-4317-8b4d-46754b24ebd0.jpg)

**Reply by squidfunk** [reply]
Nice! That looks indeed very slick 💅

**Reply by Andre601** [reply]
Still wish there was some plugin for MkDocs that would utilize `<details>` like this to generate the docs based on an OpenAPI Spec.
The ones suggested here are... meh. Mostly not really fitting that well into the style of material.

(Although, now details look more like the stuff swagger is using)

**Reply by Andre601** [reply]
Funny enough is the actual docs not updated yet... Guess it's caching doing its thing here, so I'll have to wait for that to update.

**Reply by squidfunk** [reply]
Are you referring to [our documentation](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#supported-types)? It is, in fact, using the latest admonition design.

**Reply by Andre601** [reply]
No. [My own docs](https://docs.purrbot.site)

**Comment by Andre601** [comment]
Decided to try and switch over to OpenAPI spec using the swagger-ui-tag plugin mentioned in these discussions.

Changed some CSS a bit to make colours match the general MkDocs material theme a bit closer.
![grafik](https://user-images.githubusercontent.com/11576465/232885591-fbe91f40-56ec-45f2-9991-c45c56ee3431.png)

**Comment by Hela-Masri-shift** [comment]
how such a plugin can work if we use a github action to publish the website to a github pages?

**Reply by Andre601** [reply]
It just works?
All you need is a yaml/json file of your openapi spec and the proper dependencies installed.

As of right now am I using my own hook for the api generation together with essentials-oapi to parse a json file.

**Comment by Andre601** [comment]
Just for the sake of being up-to-date and giving proper info is here my current setup:

I use a hook (Single python file MkDocs can load to execute stuff almost like a plugin) to look for a specific text in a page and when found use Essentials OpenAPI to create a new Jinja2 parser with modified templates to generate the output.

It's effectively the [mkdocsoad plugin](https://www.neoteroi.dev/mkdocs-plugins/web/oad/), but modified to fit my own design need (And to reduce depenencies).

The hook: https://github.com/purrbot-site/Docs/blob/master/theme/.hooks/api_docs_gen.py
The jinja2 files: https://github.com/purrbot-site/Docs/tree/master/theme/.hooks/theme/views_mkdocs

The result:
![grafik](https://github.com/squidfunk/mkdocs-material/assets/11576465/b06ab3c8-0f9c-486f-8214-ce7b96907a8f)
![grafik](https://github.com/squidfunk/mkdocs-material/assets/11576465/1d46e4cb-68b2-428e-b305-3ac8dda111e2)
![grafik](https://github.com/squidfunk/mkdocs-material/assets/11576465/e6c4d574-08b1-45aa-a530-ecbc0608f299)
